jpype/_pykeywords.py

```diff
@@ -16,20 +16,64 @@
 #
 # *****************************************************************************
 
+import typing
+
 # This is a superset of the keywords in Python.
 # We use this so that jpype is a bit more version independent.
-# Removing keywords from this list impacts the exposed interfaces, and therefore is a breaking change. 
-_KEYWORDS = set((
-    'False', 'None', 'True', 'and', 'as', 'assert', 'async',
-    'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else',
-    'except', 'exec', 'finally', 'for', 'from', 'global', 'if', 'import',
-    'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'print',
-    'raise', 'return', 'try', 'while', 'with', 'yield'
-))
+# Adding and removing keywords from this list impacts the exposed interfaces,
+# and therefore is technically a breaking change.
+_KEYWORDS = {
+    'False',
+    'None',
+    'True',
+    'and',
+    'as',
+    'assert',
+    'async',
+    'await',
+    'break',
+    'class',
+    'continue',
+    'def',
+    'del',
+    'elif',
+    'else',
+    'except',
+    'exec',
+    'finally',
+    'for',
+    'from',
+    'global',
+    'if',
+    'import',
+    'in',
+    'is',
+    'lambda',
+    'nonlocal',
+    'not',
+    'or',
+    'pass',
+    'print',  # No longer a Python 3 keyword. Kept for backwards compatibility.
+    'raise',
+    'return',
+    'try',
+    'while',
+    'with',
+    'yield',
+}
+
 
+def pysafe(s: str) -> typing.Optional[str]:
+    """
+    Given an identifier name in Java, return an equivalent identifier name in
+    Python that is guaranteed to not collide with the Python grammar.
 
-def pysafe(s):
-    if s.startswith("__"):
+    """
+    if s.startswith("__") and s.endswith('__'):
```

Comment:
Leading double underscores should be fine (they are protected by name mangling in Python, but not the same as "dunder" names)

Reply:
I decided that all double underscores which is a reserved name in java should not be included.  Because you could create the weird issue that a method has 2 leading and one trailing and ends up managled to add an extra.   We could just make a new mangling scheme for leading underscores, but given i found it 
 to be documemted as reserved I decided to simply remove all.

Reply:
Thanks. I have no problem rolling it back, but couldn't find the docs on the Java reserved status (I was looking on Google, and in https://docs.oracle.com/javase/specs/jls/se11/html/jls-6.html#jls-6.2). Any pointers that I can add in the comments?

Reply:
The conventions have changed through time.   I can't locate the old sun docs which irc were more explicit.  The general rule is that underscore and dollar are valid names for packages, variables, and classes, but dollar should not be used (compiler reserved) and underscore should only be used if the next character is not a valid character for starting an identifier.  So ``_2cool`` is fine because 2 can't start an identifier.   Double underscore  would fail to follow that convention and thus be discouraged.   Of course starting names with underscore is also reserved in C (for compiler and linker symbols), but since there is nothing preventing it you will find plenty of code with it. 

 https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

But no I don't see an explicit prohibition.

Reply:
I was about to update the PR with the removal of the ``endswith('__')`` but realised that it was a breaking change introduced in ``v1.4`` when compared to ``v1.3``. Instead I've added an explicit test for a Java method with a leading (but not trailing) double underscore. 

If you find this unpleasant, please let me know, and I will just remove it and re-introduce the prohibition of *any* name starting with two or more underscores which was introduced in JPype 1.4.

Reply:
Do we need to consider that case in which a "__XX_" becomes "__XX__".... I suppose it is unlikely given unless a user adds something like that to the keywords list it won't get converted to an unsafe pattern.

Reply:
I'm sorry, but I don't follow (perhaps the formatting is confusing me). 

Are you perhaps giving a case where we add the trailing ``_`` and it ends up becoming a protected name. For example:

 * ``__bad_`` gets added to the keywords
 * ``__bad_`` goes through PySafe, but because it isn't safe gets converted to ``__bad__``
 * ``__bad__`` is a protected dunder name :bomb: 

If so, I think it is quite contrived and probably not worth the effort to worry about (we will know if we are adding such a keyword in the future)